storage/node_liveness.go

```diff
@@ -0,0 +1,307 @@
+		expiration = lr.liveness.LastHeartbeat.Add(LivenessThreshold.Nanoseconds(), 0)
+	} else {
+		expiration = lr.received.Add(LivenessThreshold.Nanoseconds(), 0)
+	}
+	return nl.clock.Now().Less(expiration)
+}
+
+// NodeLiveness encapsulates information on node liveness and provides
+// an API for querying, updating, and invalidating node
+// liveness. Nodes periodically "heartbeat" the range holding the node
+// liveness system table to indicate that they're available. The
+// resulting liveness information is used to ignore unresponsive nodes
+// while making range quiescense decisions, as well as for efficient,
+// node liveness epoch-based range leader leases.
+type NodeLiveness struct {
+	clock           *hlc.Clock
+	db              *client.DB
+	gossip          *gossip.Gossip
+	manualHeartbeat chan struct{}
+	mu              struct {
+		syncutil.RWMutex
+		nodes map[roachpb.NodeID]livenessRecord
+	}
+}
+
+// NewNodeLiveness returns a new instance of NodeLiveness configured
+// with the specified gossip instance.
+func NewNodeLiveness(clock *hlc.Clock, db *client.DB, g *gossip.Gossip) *NodeLiveness {
+	nl := &NodeLiveness{
+		clock:           clock,
+		db:              db,
+		gossip:          g,
+		manualHeartbeat: make(chan struct{}, 1),
+	}
+	nl.mu.nodes = map[roachpb.NodeID]livenessRecord{}
+
+	livenessRegex := gossip.MakePrefixPattern(gossip.KeyNodeLivenessPrefix)
+	nl.gossip.RegisterCallback(livenessRegex, nl.livenessGossipUpdate)
+
+	return nl
+}
+
+// IsLive returns whether or not the specified node is considered live
+// based on the last receipt of a liveness update via gossip. It is an
+// error if the specified node is not in the local liveness table.
+func (nl *NodeLiveness) IsLive(nodeID roachpb.NodeID) (bool, error) {
+	lr, err := nl.getLiveness(nodeID)
+	if err != nil {
+		return false, err
+	}
+	return lr.isLive(nl), nil
+}
+
+// StartHeartbeat starts a periodic heartbeat to refresh this node's
+// last heartbeat in the canonical node liveness table.
+func (nl *NodeLiveness) StartHeartbeat(ctx context.Context, stopper *stop.Stopper) {
+	log.VEventf(1, ctx, "starting liveness heartbeat for node %d", nl.gossip.GetNodeID())
+	stopper.RunWorker(func() {
+		ticker := time.NewTicker(LivenessHeartbeatInterval)
+		defer ticker.Stop()
+		for {
+			if err := nl.heartbeat(ctx); err != nil {
+				log.Errorf(ctx, "failed liveness heartbeat for node %d: %s", nl.gossip.GetNodeID(), err)
+			}
+			select {
+			case <-ticker.C:
+			case <-nl.manualHeartbeat:
+			case <-stopper.ShouldStop():
+				return
+			}
+		}
+	})
+}
+
+// ManualHeartbeat triggers a heartbeat outside of the normal periodic
+// heartbeat loop. Used for unittesting.
+func (nl *NodeLiveness) ManualHeartbeat() {
+	nl.manualHeartbeat <- struct{}{}
+}
+
+// heartbeat is called to update a node's last heartbeat. This method
+// does a conditional put on the node liveness record, and if
+// successful, stores the updated liveness record in the nodes map.
```

Comment:
Why do we need to update our own liveness record here instead of waiting for it to come back to us via gossip? Do we ever even need to look up our own liveness record? Seems like we only need to do this when we're considering stealing someone else's lease. (The old liveness that we use for the CPut expected value could be local to the StartHeartbeat loop)